cache: enable Watch on arbitrary start_revision

apullo777 opened the pull request:
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

@serathius @MadhavJivrajani

serathius commented on tests/integration/cache_test.go:
```diff
@@ -197,61 +237,144 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
+	t.Log("Open watchers (prewrite = false)")
 	watches := make([]clientv3.WatchChan, len(tcs))
 	for i, tc := range tcs {
+		if tc.prewrite {
```
What is prewrite? You never mentioned it in discussions. I don't see a how this improves the test. Please remove.

apullo777 replied:
Sorry for not communicating about `prewrite`. I was thinking that perhaps we should also test scenarios where watchers are being opened after the cache has already buffered initial batch of events. So I wanted to use a flag to control when we open the watch relative to the initial batch of writes (revisions 2–10) vs. the later batch (revisions 11–12). I will remove the logic if it does not work or is unnecessary.

apullo777 replied:
Got it! I will remove the `prewrite` flag and restructure the test in this order (1) write the first batch of events, (2) open all watches with `WithRev(2)`, and then (3) write the second batch of events.

serathius replied:
(2) Please also open watches without `WithRev(2)`.

apullo777 replied:
Ah, I finally got it. Then I will open watches without `WithRev()` and add that explicit `WithRev(2)` to test cases that need to see revisions 2–10. Thanks!

serathius commented on tests/integration/cache_test.go:
```diff
@@ -197,61 +236,130 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
+	t.Log("Open watches (WithRev(2))")
+	watches := make([]clientv3.WatchChan, len(tcs))
+	for i, tc := range tcs {
+		opts := append([]clientv3.OpOption{clientv3.WithRev(2)}, tc.opts...)
```
Don't change arguments!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

serathius replied:
```suggestion
```
Don't change arguments!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

serathius commented on tests/integration/cache_test.go:
```diff
@@ -160,98 +199,160 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
+			},
+		},
+		{
+			name:       "Watch from rev 11 with single key /foo/future",
+			key:        "/foo/future",
+			opts:       []clientv3.OpOption{clientv3.WithRev(11)},
```
```suggestion
			name:       "Watch with single key /foo/future",
			key:        "/foo/future",
			opts:       nil,
```
For watcher on revisions written after we open a watch please drop WithRev

apullo777 commented on tests/integration/cache_test.go:
```diff
@@ -160,98 +199,160 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
+		{
+			name:       "Watch from rev 12 with txn prefix /foo",
+			key:        "/foo",
+			opts:       []clientv3.OpOption{clientv3.WithRev(12), clientv3.WithPrefix()},
```
This test is intended to wait until the future moment when rev == 12 before they start receiving events, so it shouldn’t receive rev == 11. Perhaps I should drop the `WithRev(11)` for "Watch from rev 11 with single key /foo/future" and rename it, and keeps `WithRev(12)` for this one?

serathius replied:
Duplicate test, have one that watches for future revision (`WithRev(11)`, `WithRev(12)`) and one that watches for latest revision where the next revision is `rev==11`.

serathius commented on tests/integration/cache_test.go:
```diff
@@ -160,98 +199,160 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
+	t.Log("Open watches")
+	watches := make([]clientv3.WatchChan, len(tcs))
+	for i, tc := range tcs {
+		watches[i] = watcher.Watch(ctx, tc.key, tc.opts...)
```
FYI, you might need to add a sleep here to ensure that next event for watches opened on `rev:0` will be `rev:11`.

serathius replied:
Maybe put the sleep after the for loop.

apullo777 commented:
Updated the PR:

-  What's changed: in `Watch()` I added a check on `startRev != 0 && startRev < demux.PeekOldest()` to reject watch that requests out-of-window revisions.
-  TODO: handling non‑zero `startRev` on an empty cache to guarantee of gap‑free replay.

serathius commented:
> handling non‑zero startRev on an empty cache to guarantee gap‑free replay. After that, KeyPredicate in Filter?

Don't understand what you mean.
